### PR TITLE
Add support for Visual Studio Team Services #53

### PR DESCRIPTION
**Pull request by Studiosixer**
Modified SrcSrv.cs to write a different form of data to the pdbs when vstsData is passed to the function.  this form includes using the SRCSRVCMD with "tf.exe git" to pull the source from Visual Studio Team Services. So when the Visual Studio Team Services provider is used, the project name and repository are extracted from the git url and this information is passed to the CreateSrcSrv function.

### CODE REVIEW

File: src/GitLink/Extensions/ProjectExtensions.cs [hunk 1]
```diff
@@ -28,15 +28,22 @@ public static string GetProjectName(this Project project)
             return projectName ?? Path.GetFileName(project.FullPath);
         }
 
-        public static void CreateSrcSrv(this Project project, string rawUrl, string revision, Dictionary<string, string> paths, bool downloadWithPowershell)
+        public static void CreateSrcSrv(this Project project, string rawUrl, string revision, Dictionary<string, string> paths, bool downloadWithPowershell, Dictionary<string, string> vstsData = null)
```
Comment (GeertvanHorrik) [inline src/GitLink/Extensions/ProjectExtensions.cs]: I am considering to move this into a SrcSrvContext class. Then we can introduce changes without breaking anything (we just have to add stuff to the SrcSrvContext class. Interested in adding this as well?

File: src/GitLink/Providers/VisualStudioTeamServicesProvider.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+﻿using System;
```
Comment (GeertvanHorrik) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: Please add file header here. If you have ReSharper, just hit CTRL E + F

File: src/GitLink/Providers/VisualStudioTeamServicesProvider.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        public VisualStudioTeamServicesProvider() 
+            : base(new GitPreparer())
+        {
+        }
+
+        public override string RawGitUrl
+        {
+            get
+            {
+                return "";
```
Comment (GeertvanHorrik) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: Please use string.Empty (easier for the brain to read, the brain now has to analyze empty string vs space).

File: src/GitLink/Providers/VisualStudioTeamServicesProvider.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            var match = _visualStudioTeamServicesRegex.Match(url);
+
+            if (!match.Success)
+            {
+                return false;
+            }
+
+            CompanyName = match.Groups["accountname"].Value;
+            CompanyUrl = match.Groups["companyurl"].Value + "DefaultCollection/";
```
Comment (GeertvanHorrik) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: Do we want "DefaultCollection" to be customizable? Or isn't that possible with team services?

File: src/GitLink/Providers/VisualStudioTeamServicesProvider.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+            ProjectName = match.Groups["project"].Value;
+            if(ProjectName == "")
```
Comment (GeertvanHorrik) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: if (string.IsNullOrWhitespace(ProjectName))

File: src/GitLink/Providers/VisualStudioTeamServicesProvider.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            }
+
+            ProjectUrl = match.Groups["companyurl"].Value + ProjectName + "/";
+
+            if (!CompanyUrl.StartsWith("https://", StringComparison.InvariantCultureIgnoreCase))
+            {
+                CompanyUrl = String.Concat("https://", CompanyUrl);
+            }
+
+            if (!ProjectUrl.StartsWith("https://", StringComparison.InvariantCultureIgnoreCase))
```
Comment (GeertvanHorrik) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: StartsWithIgnoreCase

File: src/GitLink/Providers/VisualStudioTeamServicesProvider.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+            ProjectName = match.Groups["project"].Value;
+            if(ProjectName == "")
+            {
+                ProjectName = match.Groups["repo"].Value;
+            }
+
+            ProjectUrl = match.Groups["companyurl"].Value + ProjectName + "/";
+
+            if (!CompanyUrl.StartsWith("https://", StringComparison.InvariantCultureIgnoreCase))
```
Comment (GeertvanHorrik) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: StartsWithIgnoreCase

**Comment by GeertvanHorrik** [comment]
This pr looks really good, thanks for respecting our coding standards.

File: src/GitLink/Extensions/ProjectExtensions.cs (hunk 1 above)
Comment (Studiosixer) [inline src/GitLink/Extensions/ProjectExtensions.cs]: That sounds good. I'll look into adding that.

File: src/GitLink/Providers/VisualStudioTeamServicesProvider.cs (hunk 4 above)
Comment (Studiosixer) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: It is possible to have multiple collections with TFS (VSTS on prem). With VSTS, however, looking at [this blog post](https://blogs.msdn.microsoft.com/visualstudioalm/2016/04/09/merging-the-concepts-of-account-and-collection/), it looks like we don't even need to specify a collection anymore. Each visualstudio.com account has a single "collection". So I think we can take out "DefaultCollection". But I'll do some testing to verify that.

Comment (GeertvanHorrik) [inline src/GitLink/Providers/VisualStudioTeamServicesProvider.cs]: Thanks. If this is verified, then it's all good. Or it's up to the next person to create a separate PR for this ;-)

File: src/GitLink/Pdb/SrcSrvContext.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+//   Copyright (c) 2014 - 2016 CatenaLogic. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace GitLink.Pdb
+{
+    using System.Collections.Generic;
+
+    public class SrcSrvContext
```
Comment (Studiosixer) [inline src/GitLink/Pdb/SrcSrvContext.cs]: I'm not a C# coder by trade, so let me know whether or not I'm using this class properly...

Comment (GeertvanHorrik) [inline src/GitLink/Pdb/SrcSrvContext.cs]: I think we can put all options in the context:

```
public class SrcSrvContext
{
    public SrcSrvContext()
    {
        Paths = new List<Tuple<string, string>>();
        VstsData = new Dictionary<string, string>();
        
    }

    public string Revision { get; set; }

    public List<Tuple<string, string>> Paths { get; private set; }

    public Dictionary<string, string> VstsData { get; private set; }
}
```

Then we can add stuff in the future if we have to.

Comment (Studiosixer) [inline src/GitLink/Pdb/SrcSrvContext.cs]: Ah I see. I'll work on adding this.

File: src/GitLink/Extensions/ProjectExtensions.cs [hunk 9]
```diff
@@ -28,31 +28,31 @@ public static string GetProjectName(this Project project)
             return projectName ?? Path.GetFileName(project.FullPath);
         }
 
-        public static void CreateSrcSrv(this Project project, string rawUrl, string revision, Dictionary<string, string> paths, bool downloadWithPowershell, SrcSrvContext srcSrvContext)
+        public static void CreateSrcSrv(this Project project, SrcSrvContext srcSrvContext)
         {
             Argument.IsNotNull(() => project);
```
Comment (GeertvanHorrik) [inline src/GitLink/Extensions/ProjectExtensions.cs]: Also add Argument.IsNotNull(() => srcSrvContext);

File: src/GitLink/Extensions/ProjectExtensions.cs [hunk 10]
```diff
@@ -28,31 +28,31 @@ public static string GetProjectName(this Project project)
...
             var srcsrvFile = GetOutputSrcSrvFile(project);
 
-            CreateSrcSrv(project, rawUrl, revision, paths, srcsrvFile, downloadWithPowershell, srcSrvContext);
+            CreateSrcSrv(project, srcsrvFile, srcSrvContext);
         }
 
-        public static void CreateSrcSrv(this Project project, string rawUrl, string revision, Dictionary<string, string> paths, string srcsrvFile, bool downloadWithPowershell, SrcSrvContext srcSrvContext)
+        public static void CreateSrcSrv(this Project project, string srcsrvFile, SrcSrvContext srcSrvContext)
         {
             Argument.IsNotNull(() => project);
```
Comment (GeertvanHorrik) [inline src/GitLink/Extensions/ProjectExtensions.cs]: Also add Argument.IsNotNull(() => srcSrvContext);

File: src/GitLink/Linker.cs [hunk 11]
```diff
@@ -184,6 +184,12 @@ private static bool LinkProject(Context context, Project project, string pdbStrF
                 var outputPdbFile = project.GetOutputPdbFile();
                 var projectPdbFile = pathPdbDirectory != null ? Path.Combine(pathPdbDirectory, Path.GetFileName(outputPdbFile)) : Path.GetFullPath(outputPdbFile);
                 var projectSrcSrvFile = projectPdbFile + ".srcsrv";
+
+                var srcSrvContext = new SrcSrvContext();
```
Comment (GeertvanHorrik) [inline src/GitLink/Linker.cs]: You can also use initializer code:

```
var srcSrvContext = new SrcSrvContext
{
    Revision = shaHash,
    RawUrl = context.Provider.RawGitUrl,
    DownloadWithPowershell = DownloadWithPowershell
};
```

**Comment by GeertvanHorrik** [comment]
Looks great, thanks! One last thing I always seem to forgot: you can add yourself to the CONTRIBUTORS file in the root of the repository. Then I will merge.

**Pull request merged**